File: 2021/day02/main.hs
import Data.Complex

type Cfloat = Complex Float

parseComplex :: [String] -> Cfloat
parseComplex [direction, value] =
  let x = read value :: Float
  in case direction of
       "forward" -> (x :+ 0.0)
       "down"    -> (0.0 :+ x)
       "up"      -> (0.0 :+ (-x))

reducer :: (Cfloat, Cfloat) -> Cfloat -> (Cfloat, Cfloat)
reducer (x, y) n = (x + n, y + (1 :+ (imagPart x)))

score :: Cfloat -> Int
score x = floor $realPart x * imagPart x

main :: IO()
main = do
  xs <- map (parseComplex.words) . lines <$> readFile "input"
  let (silver, gold) = foldl reducer (0, 0) xs
  print $ score silver
  print $ score gold

File: 2021/day02/main.rb
# /user/bin/env ruby
# frozen_string_literal: true

I = File.read("input")
        .scan(/(\w+) (\d+)/)
        .map do |w, x|
                case w
                when "forward" then x.to_i
                when "down"    then Complex(0, x.to_i)
                when "up"      then Complex(0, -x.to_i)
                end
        end

def prod(x) = x.rectangular.reduce(:*)

silver, gold = I.reduce([0, 0]) do |sum, x|
 sum = [sum[0] + x,
        sum[1] + x.real * Complex(1, sum[0].imag)]
end

puts "Day 02\n",
"==================\n",
"✮: #{prod(silver)}\n",
"★: #{prod(gold)}"